### PR TITLE
Initial rough draft of an Agent2Agent Transport Layer HIPE

### PR DESCRIPTION
**Pull request by swcurran**
Signed-off-by: Stephen Curran <swcurran@gmail.com>

A first cut of the Transport Layer HIPE to get the conversation started. It's a surprisingly simple set of messages - just two.  As such, it is probably too simple and I'm missing a lot. Regardless, I hope it will get the conversation going.  The document is desperately needing some pictures, but I wanted to get this out to the community for discussion.

Hack away!

### CODE REVIEW

**Comment by ryanwest6** [comment]
Overall this looks like a great start, and I agree with most of what has been said. Some initial comments:

The TDID section lost me somewhat as well. Is this TDID only to be used for inter-agency communication? How is it different from an endpoint, with its corresponding verkey and public DID?

The DID vs key issue that @dhh1128 noted seems to be something worth discussing more, since we will be dealing with multiple edge/cloud agents for each user. I don't know a lot about that but I'd love to hear more about the reasoning behind it.

Same with error handling, a discussion here of possible solutions would be good. We could potentially have the "reply-to" property as an optional attribute but I don't think it should be required. It may be necessary to assume that there was an error unless some sort of acknowledgement was received.

**Comment by swcurran** [comment]
Updated the document to add some clarifications - particularly around the layering motivation of this HIPE and the concept of TDIDs.  I left in this concept of using DIDs for "TDIDs" instead of making them simply endpoints/keys and am glad to have the discussion of the pros and cons.

**Comment by dbluhm** [comment]
[Rendered Text](https://github.com/swcurran/indy-hipe/blob/master/text/transport-layer-messaging/README.md)

**Comment by kdenhartog** [comment]
I'm hesitant to the idea that we're calling domain design implementor's specific. I think this violates almost all 7 of the Privacy by Design principles (https://www.ryerson.ca/pbdce/certification/seven-foundational-principles-of-privacy-by-design/) and conflicts with principles 2 and 10 of the Self-Sovereign Identity (https://github.com/ChristopherA/self-sovereign-identity/blob/master/self-sovereign-identity-principles.md). More specifically, I am also going to strongly challenge the notion that encapsulation is an absolute must for this protocol, but privacy is not because I don't think any of us believe it is. Am I correct that it is our jobs as protocol designers to account for both of these or have my principles completely diverged from the rest of the groups?

**Comment by swcurran** [comment]
Kyle - I don't know what you are reading in the HIPE that is in conflict with any of Privacy by Design or SSI Principles 2 and 10. Clearly I have not been able to convey what I was trying to achieve. This is where doing this via PR is ineffective.  We need a white board.  I've thought about this a lot and it might not be the right solution, but I know it doesn't violate those principles.

**Comment by swcurran** [comment]
Kyle, I'm baffled with how to respond to your comments about privacy/SSI Principles.  I don't think we want to start every document with a treatise on each aspect that makes SSI a good model.

I recommend adding the text you feel is needed in the motivation and then, (and I think, more importantly) identify specific changes to the HIPE that you feel are weak in privacy.  

My feeling is that this HIPE makes no changes to privacy or SSI principles - it simply tries to codify how different parties can use of the primitives available to independently build products. I'd even go so far as to say the proposal specifically tries to limit information exposed about an identity to only what another identity needs to know to collaborate on a shared transaction.

File: text/transport-layer-messaging/README.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+### Wire Message Format
+
+Messages sent across the wire go between Agents within the same domain or to another domain. Per the assumptions above, for messages being sent within a domain, the sender knows the TDID (Transport DID) of the receiver. If the message is being sent across domains, the public/pairwise DID to which the message is going is resolved to determine the endpoint and public key for it's domain endpoint. This makes all wire messages simple:
+
+```
+<Base64(AnonEncrypt(RecvPubKey, message))>
+```
+
+The message is sent to the endpoint of the receiving Agent. On receipt, the recipient decrypts the wire message with the transport private key associated with the endpoint and processes the message.
```
[the remark this answers is not in the document]
Comment (swcurran) [reply text/transport-layer-messaging/README.md]: In this I'm referencing the primitive that exists in Indy.  I'll update this to say other primitives could be used as they become available.

My recommendation is to put in a HIPE that documents how Indy will use the new primitives defined by the broader community - DIF, etc. If that's JWM, that's great, but it sounds like it's going to be a while before that is formalized, let alone actionable.

**Comment by dhh1128** [comment]
@kdenhartog , I think all of us do want the strong privacy guarantees. It's good to be concerned about that issue and to watch for places where we lose something important--but I don't see privacy gaps in the HIPE right now. Do you see anything specific? Right now, my privacy alarm isn't going off. Maybe we should work on resolving this concern to everybody's satisfaction at the upcoming summit at the end of July?

**Comment by kdenhartog** [comment]
After discussion on today's call I feel much more comfortable with this proposal. I do think there's things that we can do at a technical level (in the application layer) within the protocol that can help prevent cloud agent impersonation (when a cloud agent hasn't been authorized to perform some action). @nage succinctly covered this issue at the end, so I feel satisfied with the direction we're heading with this and no long feel like we're enabling privacy violations. Along with this, I think that what @mhailstone said about adding recommendations in the description would be helpful. I'll submit some examples of them to get us started.

**Comment by dhh1128** [comment]
@swcurran I have proposed some changes to your HIPE by raising a PR against your own fork. See https://github.com/swcurran/indy-hipe/pull/1

**Comment by kdenhartog** [comment]
Thanks for adding the JOSE Header format in. I see no problems with it as a first provision. I'll add the "typ" field into the JWE implementation, so we can be compliant too.

**Comment by swcurran** [comment]
Cancelling this PR - will open separate PRs for three HIPEs that fell out of this HIPE.

**Pull request closed without merging**